crates/sui-core/src/authority.rs

```diff
@@ -1178,10 +1223,21 @@ impl AuthorityState {
         let tx_digest = *certificate.digest();
         let protocol_config = epoch_store.protocol_config();
         let shared_object_refs = input_objects.filter_shared_objects();
-        let transaction_dependencies = input_objects.transaction_dependencies();
+        let receiving_objects = certificate
```

Comment:
This is where we add all of the transaction dependencies for receiving transaction arguments.

Reply:
Do we unconditionally add all receiving objects as dependency, or is there another place where we would only include the ones that we successfully attempted to receive at runtime?

Reply:
We unconditionally add all receiving objects as dependencies here. We don't add them anywhere else/there isn't any case like the one you mention.

Reply:
Would the following scenario leads to forks then:
- There exists an object O at version V1, and it's owned by address [A]. It's just a normal address-owned object.
- A certificate C can be formed trying to receive object O at version V, since there is no way to distinguish between a normal address-owned object and a receiving objet.
- A another certificate comes in and mutates O normally, and now O is at version V2.
- C comes in to execution.
- Validator V1 hasn't pruned (O, V1) yet, and will add it as dependency.
- Validator V2 has already pruned (O, V1), and won't add it as dependency.

Reply:
Very good example -- it does indeed present a fork risk.

The only way to get around this is to only record the set of actually successfully received objects in the transaction no matter the status.. and incidentally we already have the `loaded_runtime_object_metadata` which sticks around no matter what which would tell us this so we can use that to determine the actual dependencies needed.